### ISSUE TITLE
On entry to GEMM_EX  parameter number {9,12} had an illegal value

### ISSUE THREAD

**Issue by joschu** [google/jax#304]
When I try to run my jax-based training script on the GPU, it prints the following messages several hundred times but proceeds without an exception.

    ** On entry to GEMM_EX  parameter number 12 had an illegal value
    ** On entry to GEMM_EX  parameter number 9 had an illegal value

The same script doesn't print any errors when run with a CPU-only build. This error occurs with the pip build and a custom build of master.

To reproduce, run `python jax_transformer.py alice.txt` in https://github.com/joschu/jax-exp/

**Comment by mattjj** [comment]
Thanks for reporting this, John!

This is likely an XLA:GPU issue, possibly a recent one. We'll poke at it and file an internal XLA bug if necessary.

**Comment by hawkinsp** [comment]
I can reproduce this. I'll keep digging.

**Comment by hawkinsp** [comment]
I think this happens during GEMM autotuning. It may be benign because eventually XLA gives up and uses a generic GEMM, which seems to work:

Snippet from a log (with `TF_CPP_MIN_VLOG_LEVEL=2` set)

```
2019-02-01 14:41:54.525470: I external/org_tensorflow/tensorflow/compiler/xla/service/gpu/gemm_thunk.cc:314] Executing a GemmThunk
2019-02-01 14:41:54.525477: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=-1)
2019-02-01 14:41:54.525488: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=0)
2019-02-01 14:41:54.525498: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=1)
2019-02-01 14:41:54.525508: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=2)
2019-02-01 14:41:54.525520: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=3)
2019-02-01 14:41:54.525529: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=4)
 ** On entry to GEMM_EX  parameter number 9 had an illegal value
 ** On entry to GEMM_EX  parameter number 9 had an illegal value
 ** On entry to GEMM_EX  parameter number 9 had an illegal value
 ** On entry to GEMM_EX  parameter number 9 had an illegal value
 ** On entry to GEMM_EX  parameter number 9 had an illegal value
...
 ** On entry to GEMM_EX  par2019-02-01 14:41:54.525543: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpo
se, m=384, n=96, k=8192, alpha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=5)
2019-02-01 14:41:54.525553: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=6)
2019-02-01 14:41:54.525562: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=7)
2019-02-01 14:41:54.525572: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=8)
2019-02-01 14:41:54.525582: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=9)
2019-02-01 14:41:54.525592: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=10)
2019-02-01 14:41:54.525601: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=11)
2019-02-01 14:41:54.525611: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=12)
2019-02-01 14:41:54.525621: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=13)
2019-02-01 14:41:54.525633: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=14)
2019-02-01 14:41:54.525643: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=15)
2019-02-01 14:41:54.525653: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=16)
2019-02-01 14:41:54.525662: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=17)
2019-02-01 14:41:54.525672: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=99)
2019-02-01 14:41:54.525676: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:2226] DoBlasGemmWithAlgorithm returning false because algorithm 99 uses tensor ops, but the input data type is not fp16.
2019-02-01 14:41:54.525682: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=100)
2019-02-01 14:41:54.525686: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:2226] DoBlasGemmWithAlgorithm returning false because algorithm 100 uses tensor ops, but the input data type is not fp16.
2019-02-01 14:41:54.525691: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=101)
2019-02-01 14:41:54.525695: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:2226] DoBlasGemmWithAlgorithm returning false because algorithm 101 uses tensor ops, but the input data type is not fp16.
2019-02-01 14:41:54.525700: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=102)
2019-02-01 14:41:54.525704: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:2226] DoBlasGemmWithAlgorithm returning false because algorithm 102 uses tensor ops, but the input data type is not fp16.
2019-02-01 14:41:54.525710: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=103)
2019-02-01 14:41:54.525715: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:2226] DoBlasGemmWithAlgorithm returning false because algorithm 103 uses tensor ops, but the input data type is not fp16.
2019-02-01 14:41:54.525721: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:4209] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemmWithAlgorithm(transa=NoTranspose, transb=Transpose, m=384, n=96, k=8192, alp
ha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=384, computation_type=f32, algorithm=104)
2019-02-01 14:41:54.525724: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:2226] DoBlasGemmWithAlgorithm returning false because algorithm 104 uses tensor ops, but the input data type is not fp16.
2019-02-01 14:41:54.525729: I external/org_tensorflow/tensorflow/compiler/xla/util.cc:36] Internal: Unable to autotune cuBLAS gemm on stream 0x5610733f2430; none of the 25 algorithms ran successfully
2019-02-01 14:41:54.526256: I external/org_tensorflow/tensorflow/compiler/xla/util.cc:37] *** Begin stack trace ***
...
2019-02-01 14:41:54.526265: I external/org_tensorflow/tensorflow/compiler/xla/service/gpu/gemm_thunk.cc:444] Autotune on GemmThunk %dot.3438 = f32[96,384]{0,1} dot(f32[8192,96]{1,0} %bitcast.123, f32[8192,384]{1,0} %fusion.250), lhs_contracting_dims={0}, rhs_contracting_dims={0} unsuccessful.  Will use generic gemm.
2019-02-01 14:41:54.526272: I external/org_tensorflow/tensorflow/stream_executor/stream.cc:3888] [stream=0x5610733f2430,impl=0x561073379d50] Called Stream::ThenBlasGemm(transa=NoTranspose, transb=Transpose, m=96, n=384, k=8192, alpha=1, a=0x7f04dce00000, lda=96, b=0x7f04e4b78000, ldb=384, beta=0, c=0x7f051e224000, ldc=96)
2019-02-01 14:41:54.526279: I external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_blas.cc:1940] doing cuBLAS SGEMM: at=0 bt=1 m=96 n=384 k=8192 alpha=1.000000 a=0x7f04dce00000 lda=96 b=0x7f04e4b78000 ldb=384 beta=0.000000 c=0x7f051e224000 ldc=96
```

**Comment by joschu** [comment]
Oddly, I don't get that logging output when I run with  `TF_CPP_MIN_VLOG_LEVEL=2`
The code runs about 4x slower on GPU than on CPU, but that may be unrelated to this GEMM issue.
Feel free to close if you conclude that the messages are benign.

**Comment by mattjj** [comment]
I haven't looked at your code, but I'm guessing that being slower on GPU than CPU is indicative of a problem with the GPU code we're generating here and should be considered a bug. (Correct me if that's actually expected!)

If that's the case, then hopefully these warning messages will lead us to the fix!

**Comment by joschu** [comment]
Right, I don't see any reason why it shouldn't run a lot faster on the GPU.

**Comment by mattjj** [comment]
@hawkinsp did some quick profiling just now, and we have a lead on the culprit. (Later we'll add some instructions on how to run profiling, once I learn how to do it in open-source land.)

Paraphrasing Peter: the top four things in the profile are while loops, and a good guess for where those loops are coming from is the way we lower NumPy indexing. Our lowering for advanced integer indexing is likely slow on GPU because it uses XLA While loops rather than the XLA Gather HLO (which wasn't available at the time). But those While loops often don't provide any parallelism! Using the Gather HLO generates much better GPU code for indexing operations.

@hawkinsp recently surfaced Gather in JAX (#237) and used it to implement an efficient lowering of `np.take` (#306). It should be a small jump from there to using it in our lowering of NumPy advanced integer indexing.

I'll look at revising our lowering of advanced integer indexing, which was near the top of our todo list anyway, and also at getting these handy performance profiles in open-source.

**Comment by mattjj** [comment]
Yeah, the code is definitely [hitting advanced int indexing](https://github.com/joschu/jax-exp/blob/5ef438e2bf8166ad63b7bc2d6a8019b99ee84e8f/jax_transformer.py#L158-L159). So #307 (building on #237 and #306) should help a lot, and avoid those incredibly slow loops!

To check, I ran your code on a cloud P100 machine and the execution time for 10 epochs (measured to include the compile time of the first step) got better from 9f3060a: it went from 134.6 seconds to 15.7 seconds. ~8.6x speedup isn't a bad start!

```python
    from time import time
    tic = time()
    for epoch in range(10):
        print('Epoch', epoch)
        for XY in dataset_util.iterbatches(Xtr_bt, batch_size=batch_size):
            # print(''.join([codebook.idx2token(y) for y in XY[0][0]]).replace('\n','|'))
            lossval, opt_state = update(0, opt_state, XY)
            print(lossval)
    print(time() - tic)
```

This was a performance bug, and a big win without changing the original code. I bet we can make things even faster if we wanted to dig: HLO profiles will be really handy for that, but I'm still figuring out how to dump those. (We have an [old config option for it](https://github.com/google/jax/blob/fec5f47596362815a5b4aa11f2331b6a429ec639/jax/lib/xla_bridge.py#L42), but I've completely forgotten how to use it!)

Thanks for writing up this example and uncovering this bug. We haven't kicked the tires on performance enough, but fortunately the issues that come up tend to be pretty easy to fix. Examples like this one, to treat as benchmarks, are really helpful.

(Tangentially: [this line](https://github.com/joschu/jax-exp/blob/5ef438e2bf8166ad63b7bc2d6a8019b99ee84e8f/jax_transformer.py#L165) would be both faster to execute and easier to write as a `lax.broadcasted_iota`, just because it'll avoid building out the static array, but it's probably not contributing noticeably to the timing in this case.)

**Comment by joschu** [comment]
Can confirm that the code is much faster after your latest commits--amazing!
Thanks for the quick follow-up.

**Comment by mattjj** [comment]
Glad to hear it!

We've got an internal email thread with some XLA folks about these warnings; we'll update/close this issue when we figure out more about those warnings.

(Coincidentally, I just ran into Ilya and gave him an update on JAX's progress.)

**Comment by joschu** [comment]
I created a close TF translation of the code: https://github.com/joschu/jax-exp/blob/master/tf_transformer.py. 
Here are the per-iteration timings:
CPU/TF: 0.31
CPU/JAX: 0.282
GPU/TF: 0.015
GPU/Jax: 0.021
where "GPU" was a 1080Ti.
So Jax is still a bit slower on GPU, but the performance is still pretty impressive given how new it is.

**Comment by mattjj** [comment]
Awesome, thanks for sharing that! Now we should make it even faster...

**Comment by hawkinsp** [comment]
The GEMM_EX problem is now tracked at
https://github.com/tensorflow/tensorflow/issues/25761

**Comment by hawkinsp** [comment]
The GEMM_EX error should now be fixed. Note that the fix requires either building `jaxlib` from source, or waiting until we push updated `jaxlib` wheels (perhaps later this week).

I did some digging into the TF vs JAX GPU performance difference. It seems mostly be related to the runtime memory allocator on GPU, and I filed https://github.com/google/jax/issues/417 to track improving that.

Hope that helps!
